src/misc/bleeding.ts

```diff
@@ -0,0 +1,204 @@
+import { ParticleBuilder } from 'com.destroystokyo.paper';
+import { Location, Material, Particle } from 'org.bukkit';
+import { BlockFace } from 'org.bukkit.block';
+import { Waterlogged } from 'org.bukkit.block.data';
+import { Entity, EntityType, LivingEntity } from 'org.bukkit.entity';
+import { EntityDamageEvent } from 'org.bukkit.event.entity';
+import { DamageCause } from 'org.bukkit.event.entity.EntityDamageEvent';
+
+/**
+ * Adds a bleeding effect for damageable entities by randomly generating blood puddles and splashes around entity
+ *
+ * Currently implemented features:
+ *
+ * - Damage amount and health determine how much the entity bleeds
+ * - Injured entities bleed more violently when damaged again
+ * - Bleeding exponentially decreases if not injured
+ * - Armor, clothing etc... will affect bleeding amount
+ * - It is now possible to determine entity health just by looking how much they bleed
+ *
+ * @author Juffel
+ */
+
+const TIMER_DELAY = 1000; // timer delay in millis
+const BLOOD_MATERIAL = Material.DEAD_BUBBLE_CORAL_FAN; // blood material
+const PARTICLE_DATA = Material.REDSTONE_BLOCK.createBlockData(); // block data for the blood particle
+const PARTICLE_AMOUNT = 5; // amount of particles
+
+// list of living entities that don't bleed
+const ENTITY_BLACKLIST = new Set([
+  EntityType.CREEPER,
+  EntityType.SKELETON,
+  EntityType.SPIDER,
+  EntityType.CAVE_SPIDER,
+  EntityType.ENDERMAN,
+  EntityType.SLIME,
+  EntityType.SHULKER,
+  EntityType.GHAST,
+  EntityType.STRAY,
+  EntityType.VEX,
+  EntityType.TROPICAL_FISH,
+  EntityType.PUFFERFISH,
+  EntityType.SQUID,
+]);
+
+const TIME_STEP = 0.5; // control how long the bleeding continues
+const COEFFICIENT = 0.69; // control how intensely the bleeding amount decreases at each step
+
+class BleedTask {
+  private entity: Entity;
+  private time = 0.0;
+
+  amount = 0.0; // setpoint for the bleeding amount
+
+  /**
+   * Create a bleeding effect for given entity
+   * @param amount initial bleeding amount from the range of 0.0 to 1.0
+   * @param callback Exit callback
+   */
+  constructor(
+    entity: Entity,
+    amount: number,
+    callback: (error: Error) => void,
+  ) {
+    this.entity = entity;
+    this.amount = amount;
+
+    const handle = setInterval(() => {
+      try {
+        if (!this.tick()) throw undefined;
```

Comment:
Use clearInterval instead of throwing to stop the bleeding. In Java/JavaScript world using exceptions for control flow is usually considered a bad practise (unlike in e.g. Python).